res/controllers/Hercules-DJControl-Inpulse-300-script.js

```diff
@@ -185,11 +381,362 @@ DJCi300.scratchWheel = function(channel, control, value, status, _group) {
     }
 };
 
-// Bending on the jog wheel (rotating using the edge)
-DJCi300.bendWheel = function(channel, control, value, _status, _group) {
-    var interval = DJCi300._convertWheelRotation(value);
-    engine.setValue(
-        "[Channel" + channel + "]", "jog", interval * DJCi300.bendScale);
+// Helper function that calculates samples per beat
+DJCi300._samplesPerBeat = function(deck) {
+    const sampleRate = engine.getValue(`[Channel${  deck  }]`, "track_samplerate");
+    const bpm = engine.getValue(`[Channel${  deck  }]`, "local_bpm");
+    // For some reason, multiplying by 60 makes the size 1/2 as large as it's supposed to be
+    // Hence, we multiply by 120 instead
+    const secondsPerBeat = 120/bpm;
+    const samplesPerBeat = secondsPerBeat * sampleRate;
+    return samplesPerBeat;
+};
+
+// Helper function that calculates current position of play indicator in samples
+DJCi300._currentPosition = function(deck) {
+    const beatClosest = engine.getValue(`[Channel${  deck  }]`, "beat_closest");
+    let beatDistance = engine.getValue(`[Channel${  deck  }]`, "beat_distance");
+
+    // Map beatDistance so that it scales from 0 to .5, then -.5 to 0
+    beatDistance = (beatDistance > .5) ? (beatDistance - 1) : beatDistance;
+    // Adjust beatClosest and return
+    return (DJCi300._samplesPerBeat(deck) * beatDistance) + beatClosest;
+};
+
+// Mode buttons
+DJCi300.changeMode = function(channel, control, value, _status, _group) {
+    const deck = channel;
+    const oldPadMode = DJCi300.padMode[deck];
+    DJCi300.padMode[deck] = control - 15;
```

Comment:
what's the 15 here?

Reply:
Just an offset. For some reason, the MIDI codes for changing pad modes starts at 15 (0x0F), so I subtract 15 to make Mode 1 = 0.

But in hindsight, it would be more readable if I just changed the pad mode definitions at the top.